plugins/database/src/db_plugin.cpp

```diff
@@ -7903,6 +7903,26 @@ irods::error db_mod_group_op(
                       "delete from R_USER_GROUP where group_user_id = ? and user_id = ?",
                       &icss );
         if ( status != 0 ) {
+            if (CAT_SUCCESS_BUT_WITH_NO_INFO == status) {
+                // If the removal resulted in nothing happening, we need to determine whether the targeted user is a
+                // member of the targeted group. If not, a different error should be returned so that clients and rules
+                // which care to know about such failures can differentiate it from less important uses of this error.
+                std::array<char, MAX_NAME_LEN + 1> user_in_group_id{};
+                constexpr char* user_is_in_group_query =
+                    "select user_id from R_USER_GROUP where group_user_id = ? and user_id = ?";
+
+                std::vector<std::string> bind_vars;
+                bind_vars.emplace_back(groupId);
+                bind_vars.emplace_back(userId);
+
+                const auto ec = cmlGetStringValueFromSql(
+                    user_is_in_group_query, user_in_group_id.data(), MAX_NAME_LEN, bind_vars, &icss);
+                // CAT_NO_ROWS_FOUND indicates that the user is not a member of the group, so we need to change the
+                // returned error code from CAT_SUCCESS_BUT_WITH_NO_INFO to USER_NOT_IN_GROUP.
+                if (CAT_NO_ROWS_FOUND == ec) {
+                    status = USER_NOT_IN_GROUP;
+                }
+            }
```

Comment:
So, we decided to do the check based on the error code?

Still seems like it would be better / more natural if we checked to see if the user is a member of the group first. However, that would require being in a transaction to avoid TOCTOU issues.

Thoughts?

Reply:
I think that I don't think we need this check at all. `CAT_SUCCESS_BUT_WITH_NO_INFO` is returned precisely because of this situation so I think it would be fairly safe to always return `USER_NOT_IN_GROUP`. The only situation in which this error code is returned and it is not because the user is not in the group is that the query itself fails:
https://github.com/irods/irods/blob/8bf8637b33a27f42029b58a56dc507b1ffc0309d/plugins/database/src/low_level_odbc.cpp#L489-L514

So, I am okay with checking before or after if indeed we feel the need to. Please advise.

Reply:
Seeing that code helps a great deal. Seems fine to remove the code.

Let's do it.